core/modules/replica/Request.h

```diff
@@ -0,0 +1,398 @@
+     * @param allowDuplicate  - follow a previously made request if the current one duplicates it
+     */
+    Request(ServiceProvider::pointer const& serviceProvider,
+            boost::asio::io_service& io_service,
+            std::string const& type,
+            std::string const& worker,
+            int  priority,
+            bool keepTracking,
+            bool allowDuplicate);
+
+    /**
+      * This method is supposed to be provided by subclasses for additional
+      * subclass-specific actions to begin processing the request.
+      */
+    virtual void startImpl()=0;
+
+    /**
+     * Request expiration timer's handler. The expiration interval (if any)
+     * is configured via the configuraton service. When the request expires
+     * it finishes with completion status FINISHED::EXPIRED.
+     */
+    void expired(boost::system::error_code const& ec);
+
+    /**
+     * Finalize request processing (as reported by subclasses)
+     *
+     * This is supposed to be the last operation to be called by subclasses
+     * upon a completion of the request.
+     */
+    void finish(ExtendedState extendedState);
+
+    /**
+      * This method is supposed to be provided by subclasses
+      * to finalize request processing as required by the subclass.
+      */
+    virtual void finishImpl()=0;
+
+    /**
+     * This method is supposed to be provided by subclasses to handle
+     * request completion steps, such as notifying a party which initiated
+     * the request, etc.
+     */
+    virtual void notify()=0;
+
+    /**
+     * Return 'true' if the operation was aborted.
+     *
+     * USAGE NOTES:
+     *
+     *    Nomally this method is supposed to be called as the first action
+     *    witin asynchronous handlers to figure out if an on-going aynchronous
+     *    operation was cancelled for some reason. Should this be the case
+     *    the caller is supposed to quit right away. It will be up to a code
+     *    which initiated the abort to take care of putting the object into
+     *    a proper state.
+     */
+    bool isAborted(boost::system::error_code const& ec) const;
+
+    /**
+     * Ensure the object is in the deseride internal state. Throw an
+     * exception otherwise.
+     *
+     * NOTES: normally this condition should never been seen unless
+     *        there is a problem with the application implementation
+     *        or the underlying run-time system.
+     *
+     * @throws std::logic_error
+     */
+    void assertState(State desiredState) const;
+
+    /**
+     * Set the desired primary and extended state.
+     *
+     * The change of the state is done via a method to allow extra actions
+     * at this step, such as:
+     *
+     * - reporting change state in a debug stream
+     * - verifying the correctness of the state transition
+     */
+    void setState(State state,
+                  ExtendedState extendedStat);
+
+protected:

```

Comment:
this `protected` is redundant

Reply:
do you mean `private`? It is generally better (IMHO) if the data members of the base class are private...

Reply:
Agree. I have not been 100% consistent in that sense. I've modified all request classes to follow this rule.